File: better_player-0.0.84/ios/Classes/BetterPlayerHlsTokenResourceLoaderDelegate.m
#import "BetterPlayerHlsTokenResourceLoaderDelegate.h"
#import "BetterPlayerHlsTokenStore.h"

static NSString *const kHlsTokenScheme = @"hls-token";
static NSString *const kHlsTokenQueryName = @"token";
static NSString *const kHlsExpQueryName = @"exp";

@interface BetterPlayerHlsTokenResourceLoaderDelegate ()
@property(nonatomic, copy) NSString *originalScheme;
@property(nonatomic, copy, nullable) NSString *initialToken;
@property(nonatomic, copy, nullable) NSString *initialExp;
@property(nonatomic, strong) NSURLSession *session;
@end

@implementation BetterPlayerHlsTokenResourceLoaderDelegate

- (instancetype)initWithOriginalScheme:(NSString *)scheme
                                 token:(NSString *)token
                                   exp:(NSString *)exp {
    self = [super init];
    if (self) {
        _originalScheme = [scheme copy];
        _initialToken = [token copy];
        _initialExp = [exp copy];
        NSURLSessionConfiguration *config = [NSURLSessionConfiguration ephemeralSessionConfiguration];
        config.requestCachePolicy = NSURLRequestReloadIgnoringLocalAndRemoteCacheData;
        config.URLCache = nil;
        config.timeoutIntervalForRequest = 10.0;
        _session = [NSURLSession sessionWithConfiguration:config];
    }
    return self;
}

- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader
shouldWaitForLoadingOfRequestedResource:(AVAssetResourceLoadingRequest *)loadingRequest {
    NSURL *requestUrl = loadingRequest.request.URL;
    if (requestUrl == nil || requestUrl.scheme == nil) {
        return NO;
    }
    if (![requestUrl.scheme isEqualToString:kHlsTokenScheme]) {
        return NO;
    }

    NSURL *resolvedUrl = [self resolvedUrlForRequestUrl:requestUrl];
    if (resolvedUrl == nil) {
        return NO;
    }

    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:resolvedUrl];
    request.cachePolicy = NSURLRequestReloadIgnoringLocalAndRemoteCacheData;
    request.timeoutInterval = 10.0;
    AVAssetResourceLoadingDataRequest *dataRequest = loadingRequest.dataRequest;
    if (dataRequest != nil) {
        long long requestedOffset = dataRequest.requestedOffset;
        long long requestedLength = dataRequest.requestedLength;
        if (requestedOffset > 0 || requestedLength > 0) {
            NSString *rangeHeader = nil;
            if (requestedLength > 0) {
                rangeHeader = [NSString stringWithFormat:@"bytes=%lld-%lld",
                               requestedOffset, requestedOffset + requestedLength - 1];
            } else {
                rangeHeader = [NSString stringWithFormat:@"bytes=%lld-", requestedOffset];
            }
            [request setValue:rangeHeader forHTTPHeaderField:@"Range"];
        }
    }

    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request
                                                 completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error != nil) {
            [loadingRequest finishLoadingWithError:error];
            return;
        }
        NSData *responseData = data;
        NSString *pathExt = resolvedUrl.pathExtension.lowercaseString;
        BOOL isPlaylist = [pathExt isEqualToString:@"m3u8"];
        if (responseData != nil && isPlaylist) {
            NSString *playlist = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
            if (playlist != nil) {
                NSString *rewritten = [self rewritePlaylist:playlist baseUrl:resolvedUrl];
                if (rewritten != nil) {
                    responseData = [rewritten dataUsingEncoding:NSUTF8StringEncoding];
                }
            }
        }
        if (response != nil && loadingRequest.contentInformationRequest != nil) {
            NSString *contentType = response.MIMEType;
            if (isPlaylist) {
                contentType = @"application/vnd.apple.mpegurl";
            }
            loadingRequest.contentInformationRequest.contentType = contentType;
            loadingRequest.contentInformationRequest.contentLength =
                responseData.length;
            loadingRequest.contentInformationRequest.byteRangeAccessSupported = YES;
        }
        if (responseData != nil && loadingRequest.dataRequest != nil) {
            AVAssetResourceLoadingDataRequest *dataRequest = loadingRequest.dataRequest;
            long long requestedOffset = dataRequest.requestedOffset;
            long long requestedLength = dataRequest.requestedLength;
            if (requestedOffset < 0) {
                requestedOffset = 0;
            }
            long long availableLength = responseData.length - requestedOffset;
            if (availableLength < 0) {
                availableLength = 0;
            }
            long long bytesToRespond = availableLength;
            if (requestedLength > 0 && requestedLength < bytesToRespond) {
                bytesToRespond = requestedLength;
            }
            if (bytesToRespond > 0 && requestedOffset + bytesToRespond <= responseData.length) {
                NSData *chunk = [responseData subdataWithRange:NSMakeRange((NSUInteger)requestedOffset,
                                                                          (NSUInteger)bytesToRespond)];
                [dataRequest respondWithData:chunk];
            }
        }
        [loadingRequest finishLoading];
    }];
    [task resume];
    return YES;
}

- (BOOL)resourceLoader:(AVAssetResourceLoader *)resourceLoader
shouldWaitForRenewalOfRequestedResource:(AVAssetResourceRenewalRequest *)renewalRequest {
    return [self resourceLoader:resourceLoader shouldWaitForLoadingOfRequestedResource:renewalRequest];
}

- (NSURL *)resolvedUrlForRequestUrl:(NSURL *)url {
    NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
    if (components == nil) {
        return nil;
    }
    components.scheme = self.originalScheme;

    NSMutableArray<NSURLQueryItem *> *queryItems =
        components.queryItems ? [components.queryItems mutableCopy] : [NSMutableArray array];
    NSIndexSet *indicesToRemove = [queryItems indexesOfObjectsPassingTest:^BOOL(NSURLQueryItem *item, NSUInteger idx, BOOL *stop) {
        return [item.name isEqualToString:kHlsTokenQueryName] || [item.name isEqualToString:kHlsExpQueryName];
    }];
    if (indicesToRemove.count > 0) {
        [queryItems removeObjectsAtIndexes:indicesToRemove];
    }

    NSString *token = [BetterPlayerHlsTokenStore token] ?: self.initialToken;
    NSString *exp = [BetterPlayerHlsTokenStore exp] ?: self.initialExp;
    if (token.length > 0 && exp.length > 0) {
        [queryItems addObject:[NSURLQueryItem queryItemWithName:kHlsTokenQueryName value:token]];
        [queryItems addObject:[NSURLQueryItem queryItemWithName:kHlsExpQueryName value:exp]];
    }
    components.queryItems = queryItems;
    return components.URL;
}

- (NSString *)rewritePlaylist:(NSString *)playlist baseUrl:(NSURL *)baseUrl {
    NSString *token = [BetterPlayerHlsTokenStore token] ?: self.initialToken;
    NSString *exp = [BetterPlayerHlsTokenStore exp] ?: self.initialExp;
    if (token.length == 0 || exp.length == 0) {
        return playlist;
    }
    NSArray<NSString *> *lines = [playlist componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
    NSMutableArray<NSString *> *output = [NSMutableArray arrayWithCapacity:lines.count];
    for (NSString *line in lines) {
        NSString *trimmed = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        if (trimmed.length == 0) {
            [output addObject:line];
            continue;
        }
        if ([trimmed hasPrefix:@"#"]) {
            NSString *updatedTag = [self rewriteTagLine:line baseUrl:baseUrl];
            [output addObject:updatedTag ?: line];
            continue;
        }
        NSURL *url = [self resolveUrlFromString:trimmed baseUrl:baseUrl];
        if (url == nil) {
            [output addObject:line];
            continue;
        }
        BOOL isPlaylist = [[url.pathExtension lowercaseString] isEqualToString:@"m3u8"];
        NSURL *updatedUrl = [self urlByApplyingToken:url useCustomScheme:isPlaylist];
        [output addObject:updatedUrl.absoluteString ?: line];
    }
    return [output componentsJoinedByString:@"\n"];
}

- (NSURL *)resolveUrlFromString:(NSString *)value baseUrl:(NSURL *)baseUrl {
    NSURL *url = [NSURL URLWithString:value];
    if (url == nil || url.scheme == nil) {
        return [NSURL URLWithString:value relativeToURL:baseUrl].absoluteURL;
    }
    return url;
}

- (NSURL *)urlByApplyingToken:(NSURL *)url useCustomScheme:(BOOL)useCustomScheme {
    NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
    if (components == nil) {
        return url;
    }
    NSMutableArray<NSURLQueryItem *> *queryItems =
        components.queryItems ? [components.queryItems mutableCopy] : [NSMutableArray array];
    NSIndexSet *indicesToRemove = [queryItems indexesOfObjectsPassingTest:^BOOL(NSURLQueryItem *item, NSUInteger idx, BOOL *stop) {
        return [item.name isEqualToString:kHlsTokenQueryName] || [item.name isEqualToString:kHlsExpQueryName];
    }];
    if (indicesToRemove.count > 0) {
        [queryItems removeObjectsAtIndexes:indicesToRemove];
    }
    NSString *token = [BetterPlayerHlsTokenStore token] ?: self.initialToken;
    NSString *exp = [BetterPlayerHlsTokenStore exp] ?: self.initialExp;
    if (token.length > 0 && exp.length > 0) {
        [queryItems addObject:[NSURLQueryItem queryItemWithName:kHlsTokenQueryName value:token]];
        [queryItems addObject:[NSURLQueryItem queryItemWithName:kHlsExpQueryName value:exp]];
    }
    components.queryItems = queryItems;
    if (useCustomScheme) {
        components.scheme = kHlsTokenScheme;
    }
    return components.URL ?: url;
}

- (NSString *)rewriteTagLine:(NSString *)line baseUrl:(NSURL *)baseUrl {
    NSRange uriRange = [line rangeOfString:@"URI=\""];
    NSString *quote = @"\"";
    if (uriRange.location == NSNotFound) {
        uriRange = [line rangeOfString:@"URI='"];
        quote = @"'";
    }
    if (uriRange.location == NSNotFound) {
        return line;
    }
    NSUInteger start = uriRange.location + uriRange.length;
    NSRange searchRange = NSMakeRange(start, line.length - start);
    NSRange endRange = [line rangeOfString:quote options:0 range:searchRange];
    if (endRange.location == NSNotFound || endRange.location <= start) {
        return line;
    }
    NSString *uriValue = [line substringWithRange:NSMakeRange(start, endRange.location - start)];
    NSURL *url = [self resolveUrlFromString:uriValue baseUrl:baseUrl];
    if (url == nil) {
        return line;
    }
    BOOL isPlaylist = [[url.pathExtension lowercaseString] isEqualToString:@"m3u8"];
    NSURL *updatedUrl = [self urlByApplyingToken:url useCustomScheme:isPlaylist];
    if (updatedUrl.absoluteString == nil) {
        return line;
    }
    NSString *updatedLine = [line stringByReplacingCharactersInRange:NSMakeRange(start, endRange.location - start)
                                                          withString:updatedUrl.absoluteString];
    return updatedLine;
}

@end

File: ios/Runner/BetterPlayerPlus/BetterPlayerHlsTokenResourceLoaderDelegate.swift
import AVFoundation
import Foundation

final class BetterPlayerHlsTokenResourceLoaderDelegate: NSObject, AVAssetResourceLoaderDelegate {
    static let tokenScheme = "hls-token"

    private let originalScheme: String
    private let initialToken: String?
    private let initialExp: String?
    private let session: URLSession

    init(originalScheme: String, token: String?, exp: String?) {
        self.originalScheme = originalScheme
        self.initialToken = token
        self.initialExp = exp
        let config = URLSessionConfiguration.ephemeral
        config.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        config.urlCache = nil
        config.timeoutIntervalForRequest = 10.0
        self.session = URLSession(configuration: config)
        super.init()
    }

    func resourceLoader(_ resourceLoader: AVAssetResourceLoader,
                        shouldWaitForLoadingOfRequestedResource loadingRequest: AVAssetResourceLoadingRequest) -> Bool {
        guard let requestUrl = loadingRequest.request.url,
              let scheme = requestUrl.scheme,
              scheme == Self.tokenScheme else {
            return false
        }

        guard let resolvedUrl = resolvedUrl(for: requestUrl) else {
            return false
        }

        let now = Int(Date().timeIntervalSince1970)
        let currentToken = BetterPlayerHlsTokenStore.token() ?? initialToken ?? ""
        let currentExp = BetterPlayerHlsTokenStore.exp() ?? initialExp ?? ""
        // if !currentToken.isEmpty && !currentExp.isEmpty {
        //     NSLog(
        //         "BetterPlayer HLS request: now=%d exp=%@ token=%@ url=%@",
        //         now,
        //         currentExp,
        //         currentToken,
        //         resolvedUrl.absoluteString
        //     )
        // }

        let pathExt = resolvedUrl.pathExtension.lowercased()
        let isPlaylist = pathExt == "m3u8"

        // Handle both playlists and segments directly via URLSession so
        // every request uses the latest token without redirects.

        var request = URLRequest(url: resolvedUrl)
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        request.timeoutInterval = 10.0

        if let dataRequest = loadingRequest.dataRequest {
            let requestedOffset = dataRequest.requestedOffset
            let requestedLength = dataRequest.requestedLength
            if requestedOffset > 0 || requestedLength > 0 {
                let rangeHeader: String
                if requestedLength > 0 {
                    let endOffset = requestedOffset + Int64(requestedLength) - 1
                    rangeHeader = "bytes=\(requestedOffset)-\(endOffset)"
                } else {
                    rangeHeader = "bytes=\(requestedOffset)-"
                }
                request.setValue(rangeHeader, forHTTPHeaderField: "Range")
            }
        }

        var didFinish = false
        let finishLock = NSLock()
        func finishOnce(error: Error? = nil) {
            finishLock.lock()
            defer { finishLock.unlock() }
            if didFinish { return }
            didFinish = true
            if loadingRequest.isCancelled { return }
            if let error {
                loadingRequest.finishLoading(with: error)
            } else {
                loadingRequest.finishLoading()
            }
        }

        let task = session.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                NSLog("BetterPlayer HLS fetch failed: %@ url=%@", error.localizedDescription, resolvedUrl.absoluteString)
                finishOnce(error: error)
                return
            }
            if let httpResponse = response as? HTTPURLResponse {
                // NSLog(
                //     "BetterPlayer HLS fetch response: status=%d url=%@",
                //     httpResponse.statusCode,
                //     resolvedUrl.absoluteString
                // )
            }
            var responseData = data
            if let data = responseData,
               let playlist = String(data: data, encoding: .utf8),
               let self = self {
                let rewritten = self.rewritePlaylist(playlist, baseUrl: resolvedUrl)
                responseData = rewritten.data(using: .utf8)
            }
            if loadingRequest.isCancelled {
                finishOnce()
                return
            }
            if let httpResponse = response as? HTTPURLResponse {
                loadingRequest.response = httpResponse
            }
            if let response = response, let infoRequest = loadingRequest.contentInformationRequest {
                var contentType = response.mimeType
                if isPlaylist {
                    contentType = "application/vnd.apple.mpegurl"
                }
                infoRequest.contentType = contentType
                let expectedLength = (response as? HTTPURLResponse)?.expectedContentLength ?? -1
                if expectedLength > 0 {
                    infoRequest.contentLength = expectedLength
                } else {
                    infoRequest.contentLength = Int64(responseData?.count ?? 0)
                }
                infoRequest.isByteRangeAccessSupported = true
            }
            if let responseData = responseData, let dataRequest = loadingRequest.dataRequest {
                var requestedOffset = dataRequest.requestedOffset
                let requestedLength = dataRequest.requestedLength
                if requestedOffset < 0 { requestedOffset = 0 }
                var availableLength = responseData.count - Int(requestedOffset)
                if availableLength < 0 { availableLength = 0 }
                var bytesToRespond = availableLength
                if requestedLength > 0 && requestedLength < Int64(bytesToRespond) {
                    bytesToRespond = Int(requestedLength)
                }
                if loadingRequest.isCancelled {
                    finishOnce()
                    return
                }
                if bytesToRespond > 0,
                   Int(requestedOffset) + bytesToRespond <= responseData.count {
                    let chunk = responseData.subdata(in: Int(requestedOffset)..<Int(requestedOffset + Int64(bytesToRespond)))
                    dataRequest.respond(with: chunk)
                }
            }
            finishOnce()
        }
        task.resume()
        return true
    }

    func resourceLoader(_ resourceLoader: AVAssetResourceLoader,
                        shouldWaitForRenewalOfRequestedResource renewalRequest: AVAssetResourceRenewalRequest) -> Bool {
        return self.resourceLoader(resourceLoader, shouldWaitForLoadingOfRequestedResource: renewalRequest)
    }

    private func resolvedUrl(for url: URL) -> URL? {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return nil
        }
        components.scheme = originalScheme
        var queryItems = components.queryItems ?? []
        queryItems.removeAll { item in
            item.name == "token" || item.name == "exp"
        }
        let token = BetterPlayerHlsTokenStore.token() ?? initialToken
        let exp = BetterPlayerHlsTokenStore.exp() ?? initialExp
        if let token, !token.isEmpty, let exp, !exp.isEmpty {
            queryItems.append(URLQueryItem(name: "token", value: token))
            queryItems.append(URLQueryItem(name: "exp", value: exp))
        }
        components.queryItems = queryItems
        return components.url
    }

    private func rewritePlaylist(_ playlist: String, baseUrl: URL) -> String {
        let token = BetterPlayerHlsTokenStore.token() ?? initialToken
        let exp = BetterPlayerHlsTokenStore.exp() ?? initialExp
        if token?.isEmpty ?? true || exp?.isEmpty ?? true {
            return playlist
        }
        let lines = playlist.components(separatedBy: .newlines)
        var output: [String] = []
        output.reserveCapacity(lines.count)
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty {
                output.append(line)
                continue
            }
            if trimmed.hasPrefix("#") {
                output.append(rewriteTagLine(line, baseUrl: baseUrl) ?? line)
                continue
            }
            guard let url = resolveUrl(from: trimmed, baseUrl: baseUrl) else {
                output.append(line)
                continue
            }
            let isPlaylist = url.pathExtension.lowercased() == "m3u8"
            let updatedUrl = urlByApplyingToken(url, useCustomScheme: isPlaylist)
            output.append(updatedUrl.absoluteString)
        }
        return output.joined(separator: "\n")
    }

    private func resolveUrl(from value: String, baseUrl: URL) -> URL? {
        if let url = URL(string: value), url.scheme != nil {
            return url
        }
        return URL(string: value, relativeTo: baseUrl)?.absoluteURL
    }

    private func urlByApplyingToken(_ url: URL, useCustomScheme: Bool) -> URL {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return url
        }
        var queryItems = components.queryItems ?? []
        queryItems.removeAll { item in
            item.name == "token" || item.name == "exp"
        }
        let token = BetterPlayerHlsTokenStore.token() ?? initialToken
        let exp = BetterPlayerHlsTokenStore.exp() ?? initialExp
        if let token, !token.isEmpty, let exp, !exp.isEmpty {
            queryItems.append(URLQueryItem(name: "token", value: token))
            queryItems.append(URLQueryItem(name: "exp", value: exp))
        }
        components.queryItems = queryItems
        if useCustomScheme {
            components.scheme = Self.tokenScheme
        } else if components.scheme == Self.tokenScheme {
            components.scheme = originalScheme
        }
        return components.url ?? url
    }

    private func rewriteTagLine(_ line: String, baseUrl: URL) -> String? {
        let quoteCandidates = ["\"", "'"]
        var range: Range<String.Index>? = nil
        var quote: String? = nil
        for candidate in quoteCandidates {
            if let uriRange = line.range(of: "URI=\(candidate)") {
                range = uriRange
                quote = candidate
                break
            }
        }
        guard let uriRange = range, let quote = quote else {
            return nil
        }
        let startIndex = uriRange.upperBound
        guard let endRange = line.range(of: quote, range: startIndex..<line.endIndex) else {
            return nil
        }
        let uriValue = String(line[startIndex..<endRange.lowerBound])
        guard let url = resolveUrl(from: uriValue, baseUrl: baseUrl) else {
            return nil
        }
        let isPlaylist = url.pathExtension.lowercased() == "m3u8"
        let updatedUrl = urlByApplyingToken(url, useCustomScheme: isPlaylist)
        var updated = line
        updated.replaceSubrange(startIndex..<endRange.lowerBound, with: updatedUrl.absoluteString)
        return updated
    }
}
